### PR TITLE
Allow release tags to control what APKs get built

### PR DESCRIPTION
**Pull request by bluemarvin**
New build script allows us to control what gets build base on the release tag. Examples:

 `1.1.4a` -> `assembleWavevrArmRelease assembleNoapiArmRelease assembleNoapiArm64Release assembleNoapiX86Release assembleSvrArmRelease assembleSvrArm64Release assembleOculusvrArmRelease assembleOculusvrArm64Release assembleGooglevrArmRelease assembleGooglevrArm64Release`

`1.1.4b-oculusvr:arm64` -> `assembleOculusvrArm64Release`

`1.1.4c:all` - > `assembleWavevrArm assembleNoapiArm assembleNoapiArm64 assembleNoapiX86 assembleSvrArm assembleSvrArm64 assembleOculusvrArm assembleOculusvrArm64 assembleGooglevrArm assembleGooglevrArm64`

`1.1.4d-oculusvr:arm64-wavevr:arm-noapi:x86` -> `assembleOculusvrArm64Release assembleWavevrArmRelease assembleNoapiX86Release`

`1.1.4e:all-oculusvr:arm64-wavevr:arm-noapi:x86` -> `assembleOculusvrArm64 assembleWavevrArm assembleNoapiX86`

### CODE REVIEW

File: tools/taskcluster/build_targets.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+     if mode == 'r' or mode == 'release':
+        return 'Release'
+     elif mode == 'd' or mode == 'debug':
+        return 'Debug'
+     elif mode == 'a' or mode == 'all':
+        return '';
+   return 'Release'
+
+def findArch(mode):
+   archList = platforms.get(mode[0], ["arm"])
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: nit: can use single quotation marks for consistency

File: tools/taskcluster/build_targets.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+   if len(mode) > 1:
+      value = mode[1]
+      if value == 'a' or value == 'all':
+         return archList;
+      elif value.lower() in archList:
+         return [value.lower()]
+   return [archList[0]]
+
+def main(name, argv):
+   targets =['tag']
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: nit missing space after `=`

File: tools/taskcluster/build_targets.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+   for item in targets[1:]:
+      itemList = item.split(':')
+      if itemList[0] not in platforms:
+         print >> sys.stderr, "Error: " + itemList[0] + " is not a supported platform"
+         continue
+      archList = findArch(itemList)
+      for arch in archList:
+         command.append('assemble' + itemList[0].capitalize() + arch.capitalize() + mode)
+   
+   print ' '.join(command)
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: nit: can you add a code comment here (or ideally a [docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) at the top of the file) that explains the command is called via `.taskcluster.yml` and perhaps list a few examples of input -> output?

File: tools/taskcluster/build_targets.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+def findMode(arg):
+   values = arg.split(':')
+   if len(values) > 1:
+     mode = values[1].lower()
+     if mode == 'r' or mode == 'release':
+        return 'Release'
+     elif mode == 'd' or mode == 'debug':
+        return 'Debug'
+     elif mode == 'a' or mode == 'all':
+        return '';
```
[the remark this answers is not in the document]
Comment (bluemarvin) [reply tools/taskcluster/build_targets.py]: That's muscle memory from too much C/C++

File: tools/taskcluster/build_targets.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+   mode = findMode(targets[0])
+   size = len(targets)
+   if size == 1:
+      for value in platforms.keys():
+         targets.append(value + ':all')
+   command = []
+   for item in targets[1:]:
+      itemList = item.split(':')
+      if itemList[0] not in platforms:
+         print >> sys.stderr, "Error: " + itemList[0] + " is not a supported platform"
```
[the remark this answers is not in the document]
Comment (cvan) [reply tools/taskcluster/build_targets.py]: just noticed that this will leak the error messages to the [shell call](https://github.com/MozillaReality/FirefoxReality/pull/1132/files#diff-ac0229d1171c0983b27003af4c7441a5R126)

[the remark this answers is not in the document]
Comment (bluemarvin) [reply tools/taskcluster/build_targets.py]: It is sent to stderr so it will show in the taskcluster log with out messing up the print result at the end.

File: tools/taskcluster/build_targets.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+This script builds the release apks using the options passed in.
+"""
+import getopt;
+import subprocess
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: can remove this import; it's not used anymore

File: tools/taskcluster/build_targets.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+This script builds the release apks using the options passed in.
+"""
+import getopt;
```
[the remark this answers is not in the document]
Comment (cvan) [reply tools/taskcluster/build_targets.py]: nevermind: can remove this import; it's not used anymore

File: .taskcluster.yml [hunk 8]
```diff
@@ -10,7 +10,7 @@ tasks:
 #
 # Triggered whenever a pull request is opened or updated.
 #
-# - Build the app (all flavors)
+# - Builds NoAPI and GoogleVR flavors
```
Comment (cvan) [inline .taskcluster.yml]: :+1: nice work on updating all these comments

File: tools/taskcluster/build_targets.py (hunk 6 above)
Comment (cvan) [reply tools/taskcluster/build_targets.py]: think you can still remove the `import subprocess`

File: tools/taskcluster/build_targets.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+{{ event.version }} should be in the form: <tag name>:[all,release,debug]-<platform name>:[arm,arm64,x86]
+Some examples of {{ event.version }} and the resultant output from this script
+
+This is the default behaviour with no options. Only the Release build of each
+architecture for each supported platform is built:
+$ python build_targets.py 1.1.4a
+assembleWavevrArmRelease assembleNoapiArmRelease assembleNoapiArm64Release assembleNoapiX86Release assembleSvrArmRelease assembleSvrArm64Release assembleOculusvrArmRelease assembleOculusvrArm64Release assembleGooglevrArmRelease assembleGooglevrArm64Release
+
+Specifies only build the Arm64 version of the OculusVR platform:
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: nit: should `Arm` be `ARM`?

File: tools/taskcluster/build_targets.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+$ python build_targets.py 1.1.4b-oculusvr:arm64
+assembleOculusvrArm64Release
+
+Specifies all build types including Release and Debug:
+$ python build_targets.py 1.1.4c:all
+assembleWavevrArm assembleNoapiArm assembleNoapiArm64 assembleNoapiX86 assembleSvrArm assembleSvrArm64 assembleOculusvrArm assembleOculusvrArm64 assembleGooglevrArm assembleGooglevrArm64
+
+Specifies Release build of Arm64 OculusVR, Arm WaveVR, and x86 NoAPI
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: nit: missing trailing colon

File: tools/taskcluster/build_targets.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+This script configures what APKs are built using the options passed in
+from the taskcluster {{ event.version }}.
+
+{{ event.version }} should be in the form: <tag name>:[all,release,debug]-<platform name>:[arm,arm64,x86]
+Some examples of {{ event.version }} and the resultant output from this script
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: nit: do you want to end this line a period or colon?

File: tools/taskcluster/build_targets.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+This script configures what APKs are built using the options passed in
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: grammar nit: `what APKs` -> `which APKs`

File: tools/taskcluster/build_targets.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+from the taskcluster {{ event.version }}.
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: nit: capitalise: `Taskcluster`

File: tools/taskcluster/build_targets.py (hunk 10 above)
Comment (cvan) [reply tools/taskcluster/build_targets.py]: nit: `build` -> `builds`

File: tools/taskcluster/build_targets.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+Specifies all build types including Release and Debug:
+$ python build_targets.py 1.1.4c:all
+assembleWavevrArm assembleNoapiArm assembleNoapiArm64 assembleNoapiX86 assembleSvrArm assembleSvrArm64 assembleOculusvrArm assembleOculusvrArm64 assembleGooglevrArm assembleGooglevrArm64
+
+Specifies Release build of Arm64 OculusVR, Arm WaveVR, and x86 NoAPI
+$ python build_targets.py 1.1.4d-oculusvr:arm64-wavevr:arm-noapi:x86
+assembleOculusvrArm64Release assembleWavevrArmRelease assembleNoapiX86Release
+
+Specifies Release and Debug builds of Arm64 OculusVR, Arm WaveVR, and x86 NoAPI:
+$ python build_targets.py 1.1.4e:all-oculusvr:arm64-wavevr:arm-noapi:x86
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: obviously doesn't block, but it might be useful to eventually have unit tests for this file (and the overall project heh). someday in the near future. :smile:

File: tools/taskcluster/build_targets.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+   'googlevr': ['arm', 'arm64'],
+   'noapi': ['arm', 'arm64', 'x86'],
+   'svr': ['arm', 'arm64'],
+}
+
+def findMode(arg):
+   values = arg.split(':')
+   if len(values) > 1:
+     mode = values[1].lower()
+     if mode == 'r' or mode == 'release':
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: I like the verbosity of this, but if you want to save a little code, could check `if mode[0] == 'r'` instead, etc.

File: tools/taskcluster/build_targets.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+   mode = findMode(targets[0])
+   size = len(targets)
+   if size == 1:
+      for value in platforms.keys():
+         targets.append(value + ':all')
+   command = []
+   for item in targets[1:]:
+      itemList = item.split(':')
+      if itemList[0] not in platforms:
+         print >> sys.stderr, 'Error: "%s" is not a supported platform' % itemList[0]
```
Comment (cvan) [inline tools/taskcluster/build_targets.py]: in my testing, this can cause the error message to be included the [shell call](https://github.com/MozillaReality/FirefoxReality/pull/1132/files#diff-ac0229d1171c0983b27003af4c7441a5R129)

File: tools/taskcluster/build_targets.py (hunk 9 above)
Comment (bluemarvin) [reply tools/taskcluster/build_targets.py]: I matched the way it looks in Android Studio and for the gradle command line.

File: tools/taskcluster/build_targets.py (hunk 11 above)
Comment (bluemarvin) [reply tools/taskcluster/build_targets.py]: Period, It was a colon but I thought that was weird since each example has a colon too.

File: tools/taskcluster/build_targets.py (hunk 14 above)
Comment (bluemarvin) [reply tools/taskcluster/build_targets.py]: Yes. Unit tests would be nice.

File: tools/taskcluster/build_targets.py (hunk 15 above)
Comment (bluemarvin) [reply tools/taskcluster/build_targets.py]: I actually thought about dropping the single letter option to force verbosity.

File: tools/taskcluster/build_targets.py (hunk 16 above)
Comment (bluemarvin) [reply tools/taskcluster/build_targets.py]: It didn't when I tested it. Let me try again.

File: tools/taskcluster/build_targets.py (hunk 15 above)
Comment (cvan) [reply tools/taskcluster/build_targets.py]: heh, that's funny. probably fine to leave as is.

File: tools/taskcluster/build_targets.py (hunk 16 above)
Comment (cvan) [reply tools/taskcluster/build_targets.py]: nvm, works as expected:

```
echo "["`python tools/taskcluster/build_targets.py 1.1.4e:all-oculusvr:arm64-wavevr:arm-noapi:x86-foobar`"]"
```

**Comment by MortimerGoro** [comment]
LGTM
